File: src/shared/middlewares/auth_middleware.py
from django.utils.deprecation import MiddlewareMixin
from django.http import HttpRequest, HttpResponse, JsonResponse
from src.shared.services.token_verify_service import TokenVerifyService
from src.shared.exceptions.unauthorized_exception import UnauthorizedException
from src.modules.user.models import User
from src.shared.decorators.public_decorator import PUBLIC_KEY
from src.shared.middlewares.validation_middleware import VALIDATION_KEY


class AuthMiddleware(MiddlewareMixin):
    def process_request(self, request: HttpRequest):
        PUBLIC_PATHS = ["/admin", "/docs", "/openapi.json"]

        if any(request.path.startswith(path) for path in PUBLIC_PATHS):
            request.META[PUBLIC_KEY] = "true"

    def process_response(self, request: HttpRequest, response: HttpResponse):
        try:
            is_validation = VALIDATION_KEY in request.META
            if is_validation == True:
                return response

            is_public = PUBLIC_KEY in request.META
            if is_public == True:
                return response

            token = request.headers.get("Authorization", "").replace("Bearer ", "")
            payload = TokenVerifyService.execute(token)
            user = User.objects.filter(email=payload["sub"], deleted_at=None).first()

            if user is None:
                raise UnauthorizedException("Usuário inválido!")

            request.user = user

            return response

        except UnauthorizedException as e:
            return JsonResponse(
                {"status": e.status, "name": e.name, "details": e.message},
                status=e.status,
            )
        except Exception as e:
            return JsonResponse(
                {"status": 500, "name": "Internal Server Error", "details": str(e)},
                status=500,
            )

File: src/shared/middlewares/auth.middleware.ts
import { NextFunction, Response } from 'express';
import { IRequest } from '../domain/interfaces/request.interface';
import { UnauthorizedException } from '../exceptions/unauthorized.exception';
import { tokenVerifyService } from '../services/token-verify.service';
import { UserRequest } from '../types/user-request';
import { generateJsonError } from '../utils/error';

export const authMiddleware = async (
  request: IRequest,
  response: Response,
  next: NextFunction,
): Promise<void> => {
  try {
    if (
      request.path.startsWith('/auth') ||
      request.path.startsWith('/files') ||
      request.path.startsWith('/doc')
    ) {
      return next();
    }

    const token: string | undefined = request.headers.authorization;
    if (!token) {
      throw new UnauthorizedException('Token Inválido!');
    }

    const user: UserRequest = await tokenVerifyService.execute(token);
    if (!user) {
      throw new UnauthorizedException('Usuário inválido!');
    }

    request.user = user;

    next();
  } catch (error) {
    generateJsonError(error, response);
  }
};
